File: internal/wikidata/format.go
package wikidata

import (
	"fmt"
	"sort"
	"strings"
)

func Stringify(value any) string {
	switch typed := value.(type) {
	case nil:
		return ""
	case map[string]any:
		if values, ok := typed["values"]; ok {
			items, ok := values.([]any)
			if ok {
				formatted := make([]string, 0, len(items))
				for _, item := range items {
					itemMap, _ := item.(map[string]any)
					formatted = append(formatted, Stringify(itemMap["value"]))
				}
				return strings.Join(formatted, ", ")
			}
		}
		if raw, ok := typed["value"]; ok {
			return Stringify(raw)
		}
		if raw, ok := typed["string"]; ok {
			return Stringify(raw)
		}
		if qid, ok := typed["QID"].(string); ok && qid != "" {
			label, _ := typed["label"].(string)
			if strings.TrimSpace(label) == "" {
				return qid
			}
			return fmt.Sprintf("%s (%s)", label, qid)
		}
		if pid, ok := typed["PID"].(string); ok && pid != "" {
			label, _ := typed["label"].(string)
			if strings.TrimSpace(label) == "" {
				return pid
			}
			return fmt.Sprintf("%s (%s)", label, pid)
		}
		if amount, ok := typed["amount"]; ok {
			unit, _ := typed["unit"].(string)
			text := fmt.Sprintf("%v", amount)
			if strings.TrimSpace(unit) != "" {
				text += " " + unit
			}
			return strings.TrimSpace(text)
		}
		// Stable fallback for unknown objects.
		keys := make([]string, 0, len(typed))
		for key := range typed {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		parts := make([]string, 0, len(keys))
		for _, key := range keys {
			parts = append(parts, fmt.Sprintf("%s=%s", key, Stringify(typed[key])))
		}
		return strings.Join(parts, ", ")
	case string:
		return typed
	case fmt.Stringer:
		return typed.String()
	default:
		return fmt.Sprintf("%v", typed)
	}
}

func stringifyQualifier(q textifierQualifier) string {
	values := make([]string, 0, len(q.Values))
	for _, v := range q.Values {
		values = append(values, Stringify(v.Value))
	}
	return strings.Join(values, ", ")
}

func TripletValuesToString(entityID string, propertyID string, entity textifierEntity) string {
	if len(entity.Claims) == 0 {
		return ""
	}

	var b strings.Builder
	for claimIndex, claim := range entity.Claims {
		for valueIndex, claimValue := range claim.Values {
			if claimIndex > 0 || valueIndex > 0 {
				b.WriteString("\n")
			}

			property := claim.PID
			if strings.TrimSpace(property) == "" {
				property = propertyID
			}

			entityLabel := strings.TrimSpace(entity.Label)
			if entityLabel == "" {
				entityLabel = entityID
			}

			b.WriteString(entityLabel)
			b.WriteString(" (")
			b.WriteString(entityID)
			b.WriteString("): ")
			b.WriteString(claim.PropertyLabel)
			b.WriteString(" (")
			b.WriteString(property)
			b.WriteString("): ")
			b.WriteString(Stringify(claimValue.Value))
			b.WriteString("\n")

			rank := strings.TrimSpace(claimValue.Rank)
			if rank == "" {
				rank = "normal"
			}
			b.WriteString("  Rank: ")
			b.WriteString(rank)
			b.WriteString("\n")

			if len(claimValue.Qualifiers) > 0 {
				b.WriteString("  Qualifier:\n")
				for _, qualifier := range claimValue.Qualifiers {
					b.WriteString("    - ")
					b.WriteString(qualifier.PropertyLabel)
					b.WriteString(" (")
					b.WriteString(qualifier.PID)
					b.WriteString("): ")
					b.WriteString(stringifyQualifier(qualifier))
					b.WriteString("\n")
				}
			}

			if len(claimValue.References) > 0 {
				for referenceIndex, referenceClaims := range claimValue.References {
					b.WriteString("  Reference ")
					b.WriteString(fmt.Sprintf("%d", referenceIndex+1))
					b.WriteString(":\n")
					for _, refClaim := range referenceClaims {
						b.WriteString("    - ")
						b.WriteString(refClaim.PropertyLabel)
						b.WriteString(" (")
						b.WriteString(refClaim.PID)
						b.WriteString("): ")
						b.WriteString(stringifyQualifier(refClaim))
						b.WriteString("\n")
					}
				}
			}
		}
	}

	return strings.TrimSpace(b.String())
}

File: src/format.ts
import type {
  TextifierClaim,
  TextifierEntity,
  TextifierQualifier,
  TextifierQualifierValue,
} from "./types.js";

function asRecord(value: unknown): Record<string, unknown> | undefined {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return undefined;
  }
  return value as Record<string, unknown>;
}

export function stringify(value: unknown): string {
  if (value === null || value === undefined) {
    return "";
  }

  if (typeof value === "string") {
    return value;
  }

  if (typeof value === "object") {
    const typed = asRecord(value);
    if (!typed) {
      return String(value);
    }

    const values = typed.values;
    if (Array.isArray(values)) {
      const formatted = values.map((item) => {
        const itemMap = asRecord(item);
        return stringify(itemMap?.value);
      });
      return formatted.join(", ");
    }

    if (Object.hasOwn(typed, "value")) {
      return stringify(typed.value);
    }

    if (Object.hasOwn(typed, "string")) {
      return stringify(typed.string);
    }

    if (typeof typed.QID === "string" && typed.QID !== "") {
      const label = typeof typed.label === "string" ? typed.label : "";
      if (label.trim() === "") {
        return typed.QID;
      }
      return `${label} (${typed.QID})`;
    }

    if (typeof typed.PID === "string" && typed.PID !== "") {
      const label = typeof typed.label === "string" ? typed.label : "";
      if (label.trim() === "") {
        return typed.PID;
      }
      return `${label} (${typed.PID})`;
    }

    if (Object.hasOwn(typed, "amount")) {
      const amount = String(typed.amount);
      const unit = typeof typed.unit === "string" ? typed.unit : "";
      const text = unit.trim() === "" ? amount : `${amount} ${unit}`;
      return text.trim();
    }

    const keys = Object.keys(typed).sort();
    const parts = keys.map((key) => `${key}=${stringify(typed[key])}`);
    return parts.join(", ");
  }

  return String(value);
}

function stringifyQualifier(q: TextifierQualifier): string {
  const values: string[] = [];
  for (const value of q.values || []) {
    values.push(stringify((value as TextifierQualifierValue).value));
  }
  return values.join(", ");
}

export function tripletValuesToString(
  entityID: string,
  propertyID: string,
  entity: TextifierEntity,
): string {
  if (!entity.claims || entity.claims.length === 0) {
    return "";
  }

  const blocks: string[] = [];

  entity.claims.forEach((claim: TextifierClaim) => {
    (claim.values || []).forEach((claimValue) => {
      const lines: string[] = [];
      const property = (claim.PID || "").trim() === "" ? propertyID : claim.PID || "";
      const entityLabel =
        (entity.label || "").trim() === "" ? entityID : entity.label || entityID;

      lines.push(
        `${entityLabel} (${entityID}): ${claim.property_label} (${property}): ${stringify(claimValue.value)}`,
      );

      const rank = (claimValue.rank || "").trim() === "" ? "normal" : claimValue.rank;
      lines.push(`  Rank: ${rank}`);

      if (claimValue.qualifiers && claimValue.qualifiers.length > 0) {
        lines.push("  Qualifier:");
        for (const qualifier of claimValue.qualifiers) {
          lines.push(
            `    - ${qualifier.property_label} (${qualifier.PID}): ${stringifyQualifier(qualifier)}`,
          );
        }
      }

      if (claimValue.references && claimValue.references.length > 0) {
        claimValue.references.forEach((referenceClaims, referenceIndex) => {
          lines.push(`  Reference ${referenceIndex + 1}:`);
          for (const refClaim of referenceClaims) {
            lines.push(
              `    - ${refClaim.property_label} (${refClaim.PID}): ${stringifyQualifier(refClaim)}`,
            );
          }
        });
      }

      blocks.push(lines.join("\n"));
    });
  });

  return blocks.join("\n").trim();
}
